### PR TITLE
[STORM-1416] Documentation for state store

### PR DESCRIPTION
**Pull request by arunmahadevan**
Documentation for the state store support added to storm core as a part
of the windowing in STORM-1175.

### CODE REVIEW

**Comment by ptgoetz** [comment]
+1

File: docs/documentation/State-checkpointing.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+Storm relies on the acking mechanism to replay tuples in case of failures. It is possible that the state is committed
+but the worker crashes before acking the tuples. In this case the tuples are replayed causing duplicate state updates.
+Also currently the StatefulBoltExecutor continues to process the tuples from a stream after it has received a checkpoint
+tuple on one stream while waiting for checkpoint to arrive on other input streams for saving the state. This can also cause
+duplicate state updates during recovery.
+
+The state abstraction does not eliminate duplicate evaluations and currently provides only at-least once guarantee.
+
+### IStateful bolt hooks
+IStatefult bolt interface provides hook methods where in the stateful bolts could implement some custom actions.
```
Comment (HeartSaVioR) [inline docs/documentation/State-checkpointing.md]: IStatefult -> IStateful

File: docs/documentation/State-checkpointing.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+and is useful if the bolt manages the state on its own. This is currently used only by the internal system bolts,
+for e.g. the CheckpointSpout to save its state.
+
+`KeyValueState` implementation should also implement the methods defined in the `backtype.storm.state.KeyValueState` interface.
+
+### State provider
+The framework instantiates the state via the corresponding `StateProvider` implementation. A custom state should also provide
+a `StateProvider` implementation which can load and return the state based on the namespace. Each state belongs to a unique namespace.
+The namespace is typically unique per task so that each task can have its own state. The StateProvider and the corresponding
+State implementation should be available in the class path of the storm (by placing them in the extlib directory).
```
Comment (HeartSaVioR) [inline docs/documentation/State-checkpointing.md]: storm -> Storm (or Apache Storm)

**Comment by HeartSaVioR** [comment]
+1 Nice work!

**Comment by satishd** [comment]
+1

**Comment by revans2** [comment]
+1

**Pull request merged**